src/utils/events/wheel-input.js

```diff
@@ -0,0 +1,133 @@
+/* global window:false */
+const ua = typeof window.navigator !== 'undefined' ?
+  window.navigator.userAgent.toLowerCase() : '';
+const firefox = ua.indexOf('firefox') !== -1;
+
+const WHEEL_EVENTS = [
+  // Chrome, Safari
+  'wheel',
+  // IE
+  'mousewheel',
+  // legacy Firefox
+  'DOMMouseScroll'
+];
+
+// Constants for normalizing input delta
+const WHEEL_DELTA_MAGIC_SCALER = 4.000244140625;
+const WHEEL_DELTA_PER_LINE = 40;
+const TRACKPAD_MAX_DELTA = 4;
+const TRACKPAD_MAX_DELTA_PER_TIME = 200;
+// Slow down zoom if shift key is held for more precise zooming
+const SHIFT_MULTIPLIER = 0.25;
+
+export default class WheelInput {
+
+  constructor(element, callback) {
+    this.element = element;
+    this.callback = callback;
+
+    this.handler = this.handler.bind(this);
+
+    WHEEL_EVENTS.forEach(eventName => element.addEventListener(eventName, this.handler));
+
+    this._state = {
```

Comment:
TBH I just copied this code from `react-map-gl` and haven't really given it much thought. @Pessimistress any thoughts on @ibgreen's suggestions?